posthog/temporal/batch_exports/spmc.py

```diff
@@ -962,6 +962,80 @@ async def produce_batch_export_record_batches_from_range(
                     raise
 
 
+class RetryOOMRecorBatchAsyncIterator:
+    def __init__(
+        self,
+        clickhouse_client: ClickHouseClient,
+        query: str,
+        query_parameters: dict[str, str],
+        interval_start: dt.datetime | None,
+        interval_end: dt.datetime,
+    ):
+        self.client = clickhouse_client
+        self.query = query
+        self.query_parameters = query_parameters
+        self.interval_start = interval_start
+        self.interval_end = interval_end
+        self.streamers = []
+
+    def __aiter__(self) -> collections.abc.AsyncIterator[pa.RecordBatch]:
+        return self.stream()
+
+    async def stream(self) -> collections.abc.AsyncIterator[pa.RecordBatch]:
+        if not self.streamers and self.interval_start is not None:
+            new_streamer = self.start_new_stream(interval_start=self.interval_start, interval_end=self.interval_end)
+            self.streamers.append((new_streamer, self.interval_start, self.interval_end))
+        else:
+            # TODO: How to split up batch exports with no start date?
+            # For now, do nothing and pass along the values.
+            async for record_batch in self.start_new_stream(
+                interval_start=self.interval_start, interval_end=self.interval_end
+            ):
+                yield record_batch
+            return
+
+        while self.streamers:
+            try:
+                streamer, interval_start, interval_end = self.streamers.pop()
+            except IndexError:
+                return
+
+            while True:
+                try:
+                    yield await anext(streamer)
+
+                except ClickHouseMemoryLimitExceededError:
```

Comment:
might be worth adding some logging just so we know when this has happened and in case we need to debug